tests/end-to-end/upgrade/pkg/tests/monitoring/targets_rules.go

```diff
@@ -0,0 +1,277 @@
+	if err != nil {
+		return errors.Wrap(err, "cannot get number of nodes")
+	}
+	for {
+		actualAlertManagers := 0
+		actualPrometheusInstances := 0
+		actualNodeExporter := 0
+		actualKubeStateMetrics := 0
+		actualGrafanaInstance := 0
+		select {
+		case <-timeout:
+			tick.Stop()
+			if expectedAlertManagers != actualAlertManagers {
+				return errors.Errorf("timed out: expected alertmanager running is %d but got %d instances", expectedAlertManagers, actualAlertManagers)
+			}
+			if expectedNodeExporter != actualNodeExporter {
+				return errors.Errorf("timed out: expected node exporter running is %d but got %d instances", expectedNodeExporter, actualNodeExporter)
+			}
+			if expectedPrometheusInstances != actualPrometheusInstances {
+				return errors.Errorf("timed out: expected prometheus running is %d but got %d instances", expectedPrometheusInstances, actualPrometheusInstances)
+			}
+			if expectedKubeStateMetrics != actualKubeStateMetrics {
+				return errors.Errorf("timed out: expected kube-state-metrics running is %d but got %d instances", expectedKubeStateMetrics, actualKubeStateMetrics)
+			}
+			if expectedGrafanaInstance != actualGrafanaInstance {
+				return errors.Errorf("timed out: expected grafana running is %d but got %d instances", expectedGrafanaInstance, actualGrafanaInstance)
+			}
+
+		case <-tick.C:
+			pods, err := t.k8sCli.CoreV1().Pods(namespace).List(metav1.ListOptions{LabelSelector: "app in (alertmanager,prometheus,grafana,prometheus-node-exporter)"})
+			if err != nil {
+				return errors.Wrapf(err, "error while kubectl get pods")
+			}
+
+			for _, pod := range pods.Items {
+				podName := pod.Name
+				isReady := getPodStatus(pod)
+				if isReady {
+					switch true {
+					case strings.Contains(podName, "alertmanager"):
+						actualAlertManagers++
+
+					case strings.Contains(podName, "node-exporter"):
+						actualNodeExporter++
+
+					case strings.Contains(podName, "prometheus-monitoring"):
+						actualPrometheusInstances++
+
+					case strings.Contains(podName, "grafana"):
+						actualGrafanaInstance++
+					}
+				}
+			}
+
+			pods, err = t.k8sCli.CoreV1().Pods(namespace).List(metav1.ListOptions{LabelSelector: "app.kubernetes.io/name=kube-state-metrics"})
+			if err != nil {
+				return errors.Wrapf(err, "error while kubectl get pods")
+			}
+
+			for _, pod := range pods.Items {
+				podName := pod.Name
+				isReady := getPodStatus(pod)
+				if isReady && strings.Contains(podName, "kube-state-metrics") {
+					actualKubeStateMetrics++
+				}
+			}
+
+			if expectedAlertManagers == actualAlertManagers && expectedNodeExporter == actualNodeExporter && expectedPrometheusInstances == actualPrometheusInstances && expectedKubeStateMetrics == actualKubeStateMetrics && expectedGrafanaInstance == actualGrafanaInstance {
+				return nil
+			}
+		}
+	}
+}
+
+func (t TargetsAndRulesTest) testTargetsAreHealthy() error {
+	timeout := time.After(3 * time.Minute)
+	tick := time.NewTicker(30 * time.Second)
```

Comment:
do we need to wait 30 seconds for the first check?

Reply:
@piotrmiskiewicz the idea is that the scraping interval is 30 seconds, so if in the first tick one of the targets was down, then there is no meaning to check again in less than 30 seconds beacause promtheus would have not yet done another scraping and the check will still fail

Reply:
but at the same time you are right, this makes the first check only happens after 30 seconds, so in total that makes the test runs in around 1 minute

Reply:
so please do the first check without waiting 30 seconds